### PR TITLE
RUL-28: Add calculate action

### PR DESCRIPTION
**Pull request by mmetayer**
**Description**

**Definition Of Done**

| Q                                 | A
| --------------------------------- | ---
| Technical Review and 2 GTM        | Todo
| English Review and 1 GTM          | Todo

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 1]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+- field: the attribute code (required)
+- locale: locale code if the attribute is localizable (optional)
+- scope: channel code if the attribute is scopable (optional)
+- currency: currency code if the attribute is a price collection (optional)
+- unit: unit code if the attribute is a measurement (optional, the default measurement unit of the attribute will be used if not set)
+
+**source**: the first operand of the operation. It can be composed of either:
+
+- value: a constant value (number, for instance 3.14)
```
Comment (pierallard) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:573]: (required ?)

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 2]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+It is exactly the same format as the ``source`` property, plus one required field:
+
+- operator: can be either ``add``, ``substract``, ``multiply`` or ``divide`` (required)
+
+    .. warning::
+
+       The operations are applied in the order they are provided, regardless of any mathematical priority.
+       For instance, 5 - 3 + 2 x 5 will result in ((5 - 3) + 2) x 5) = 20
+
+If a product value required in an operation is empty, or if a division by zero occurs, the product won't be updated.
```
Comment (pierallard) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:593]: cc Morgane

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 3]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+.. tip::
+
+    For instance, in order to calculate the volume of a cone, given a radius and a height, you can use the following action:
```
Comment (pierallard) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:597]: You should write the mathematical action here, like volumne = PI * radius * radius * height / 3

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 4]
```diff
@@ -551,6 +551,96 @@ The expected values are:
             - type: clear
               field: associations
 
+Calculate
+_________
+
+This action allows to calculate numeric attribute values, with simple mathematical operations (addition, substraction, multiplication and division)
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:557]:
```suggestion
This action allows calculating numeric attribute values, with simple mathematical operations (addition, subtraction, multiplication, and division)
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 5]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+This action only accepts number, measurement or price collection attributes for both source and destination.
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:559]:
```suggestion
This action requires number, measurement or price collection attributes for both sources and destinations.
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 6]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+The action is split in 3 distinct parts:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:561]:
```suggestion
The action is split into 3 different parts:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 7]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+**destination**: the value you want to update. It can be composed of :
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:563]:
```suggestion
**destination**: the value you want to update. It is composed of:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 1 above)
Comment (camillebourjade) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:573]: unclear about the 'or' here

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 8]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+- value: a constant value (number, for instance 3.14)
+
+or:
+
+- field: attribute code of the source value (required)
+- locale: locale code if the attribute is localizable (optional)
+- scope: channel code if the attribute is scopable (optional)
+- currency: currency code if the attribute is a price collection (optional)
+
+**operation_list**: the list of operations to execute (obviously, at least one operation is required)
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:582]:
```suggestion
**operation_list**: the list of operations to execute (at least one operation is required)
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 9]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+It is exactly the same format as the ``source`` property, plus one required field:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:584]:
```suggestion
It is exactly the same format as the ``source`` property, with an additional required field:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 10]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+- operator: can be either ``add``, ``substract``, ``multiply`` or ``divide`` (required)
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:586]:
```suggestion
- operator: can be either ``add``, ``subtract``, ``multiply`` or ``divide`` (required)
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 11]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+                    - operator: multiply
+                      field: radius
+                    - operator: multiply
+                      field: radius
+                    - operator: multiply
+                      field: height
+                    - operator: divide
+                      value: 3
+
+    The following action will calculate a price in euro, based on the price in dollars and a ratio:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:618]:
```suggestion
    The following action will calculate a price in euros, based on the price in dollars and a ratio:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 12]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+                    field: price
+                    scope: ecommerce
+                    currency: USD
+                  operation_list:
+                    - operator: multiply
+                      field: ratio
+
+    .. warning::
+
+       When using measurement attributes, their value will be converted to the default measuremnt unit defined for the attribute.
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:637]:
```suggestion
       When using measurement attributes, their value will be converted to the default measurement unit defined for the attribute.
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 13]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+       For instance, given a `Length` attribute with a 'CENTIMETER' default unit, and its value for the product is 1 INCH, the value will be converted to 2.54 (1 inch = 2.54 cm)
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:638]: unclear

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 2 above)
Comment (mmetayer) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:593]: we already talked about that, this behavior was validated

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 5 above)
Comment (mmetayer) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:559]: but there is only 1 source and 1 destination. Maybe rather:
`This action requires number, measurement or price collection attributes for both the source and the destination.` ?

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 1 above)
Comment (mmetayer) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:573]: @camillebourjade it means that the `source` can be either:
```yaml
source:
  value: 3.14
```
or
```yaml
source:
  field: price
  locale: en_US
  scope: ecommerce
  currency: EUR
```
(you cannot have both `field` and `value` in the definition)

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 13 above)
Comment (mmetayer) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:638]: maybe:
`For instance, given a `Length` attribute with a 'CENTIMETER' default unit, and its value for the product is 1 INCH, the value applied in the operation will be 2.54 (1 inch = 2.54 cm)` ?

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 14]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+**destination**: the value you want to update. It is composed of:
+
+- field: the attribute code (required)
+- locale: locale code if the attribute is localizable (optional)
+- scope: channel code if the attribute is scopable (optional)
+- currency: currency code if the attribute is a price collection (optional)
+- unit: unit code if the attribute is a measurement (optional, the default measurement unit of the attribute will be used if not set)
+
+**source**: the first operand of the operation. It can be composed of either:
```
Comment (camillebourjade) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:571]:
```suggestion
**source**: the first operand of the operation. It requires at least either a value or a field and additional items. For instance, you can have:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst (hunk 5 above)
Comment (camillebourjade) [reply manipulate_pim_data/rule/general_information_on_rule_format.rst:559]: ok

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 15]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+- value: a constant value (number, for instance 3.14)
```
Comment (mmetayer) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:573]:
```suggestion
- value: a constant numeric value, e.g: 3.14 (required)
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 16]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+
+or:
```
Comment (mmetayer) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:575]:
```suggestion
OR:
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 17]
```diff
@@ -551,6 +551,96 @@ The expected values are:
             - type: clear
               field: associations
 
+Calculate
+_________
+
+This action allows calculating numeric attribute values, with simple mathematical operations (addition, subtraction, multiplication, and division)
+
+This action only accepts number, measurement or price collection attributes for both source and destination.
```
Comment (mmetayer) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:559]:
```suggestion
This action only accepts number, measurement or price collection attributes for both the source and the destination.
```

File: manipulate_pim_data/rule/general_information_on_rule_format.rst [hunk 18]
```diff
@@ -551,6 +551,96 @@ The expected values are:
...
+This action allows calculating numeric attribute values, with simple mathematical operations (addition, subtraction, multiplication, and division)
+
+This action only accepts number, measurement or price collection attributes for both the source and the destination.
+
+The action is split into 3 different parts:
+
+**destination**: the value you want to update. It is composed of:
+
+- field: the attribute code (required)
+- locale: locale code if the attribute is localizable (optional)
```
Comment (morganen) [inline manipulate_pim_data/rule/general_information_on_rule_format.rst:566]: peut être ajouter que sinon, il faut mettre "null" ?

**Pull request merged**